### PR TITLE
Fixed #28048 -- Allowed generic date views to use related fields as date_field.

### PR DESCRIPTION
**Pull request by lefterisnik**
This PR tries to resolve a small issue on the date generic views. On the date generic views right now you cannot specify as a date_field a related field (you get FieldDoesNotExist).

### CODE REVIEW

File: tests/generic_views/urls.py [hunk 1]
```diff
@@ -142,7 +142,6 @@
     url(r'^dates/books/sortedbynamedec/$',
         views.BookArchive.as_view(ordering='-name')),
 
-
```
Comment (timgraham) [inline tests/generic_views/urls.py]: Please don't make unrelated whitespace changes.

File: django/views/generic/dates.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import datetime
 
 from django.conf import settings
+from django.contrib.admin.utils import get_fields_from_path
```
Comment (timgraham) [inline django/views/generic/dates.py]: A dependency on a contrib app here isn't appropriate.

File: tests/generic_views/test_dates.py [hunk 3]
```diff
@@ -21,10 +21,16 @@ class TestDataMixin:
 
     @classmethod
     def setUpTestData(cls):
+        cls.series1 = Series.objects.create(
+            name='Django books', pubdate=datetime.date(2003, 10, 1)
```
Comment (timgraham) [inline tests/generic_views/test_dates.py]: Please including a trailing comma in cases like this so if more items are added later, we don't have to modify this line again.

File: tests/generic_views/test_dates.py [hunk 4]
```diff
@@ -21,10 +21,16 @@ class TestDataMixin:
...
+        )
         cls.artist1 = Artist.objects.create(name='Rene Magritte')
         cls.author1 = Author.objects.create(name='Roberto Bolaño', slug='roberto-bolano')
         cls.author2 = Author.objects.create(name='Scott Rosenberg', slug='scott-rosenberg')
-        cls.book1 = Book.objects.create(name='2066', slug='2066', pages=800, pubdate=datetime.date(2008, 10, 1))
+        cls.book1 = Book.objects.create(
+            name='2066', slug='2066', pages=800, series=cls.series1,
+            pubdate=datetime.date(2008, 10, 1)
```
Comment (timgraham) [inline tests/generic_views/test_dates.py]: Add trailing comma.

File: django/views/generic/dates.py (hunk 2 above)
Comment (lefterisnik) [inline django/views/generic/dates.py]: @timgraham, thank you for the comment. What do you propose? Reimplementing the `get_fields_from_path` inside the `dates.py` for example? Another idea is to extend the Meta options class API?

Comment (timgraham) [inline django/views/generic/dates.py]: Yes, this feels like a meta thing. I'm a bit surprised if something similar doesn't already exist but perhaps there are other helps that could also be consolidated.

Comment (lefterisnik) [inline django/views/generic/dates.py]: Hi @timgraham, I searched on the project but I couldn't find any helper function that I can use to get those fields (there are some functions under `django.db.models.sql.query.py` and `django.db.models.query.py`) so I added these lines: https://github.com/django/django/pull/8307/commits/2f7aedc832f44dc8d5e5bfe7953a3ecdaebb2144. Also, I don't think that it would be good idea to extend `Meta` options class because those fields are accessible as nested attributes of its  relation fields. 

Please, let me know your opinion.

**Comment by lefterisnik** [comment]
Hi @timgraham,

Could you have a look on this. I've updated the PR.

Thanks,
Lefteris Nikoltsios

File: tests/generic_views/models.py [hunk 5]
```diff
@@ -43,6 +51,7 @@ class Book(models.Model):
     slug = models.SlugField()
     pages = models.IntegerField()
     authors = models.ManyToManyField(Author)
+    series = models.ForeignKey(Series, null=True, on_delete=models.CASCADE)
```
Comment (timgraham) [inline tests/generic_views/models.py]: To follow the style of other models: `(Series, models.CASCADE null=True)`

File: tests/generic_views/test_dates.py [hunk 6]
```diff
@@ -358,6 +368,10 @@ def test_month_view_invalid_pattern(self):
         res = self.client.get('/dates/books/2007/no_month/')
         self.assertEqual(res.status_code, 404)
 
+    def test_month_view_based_on_related_fields(self):
+        res = self.client.get('/dates/books/2003/oct/related_field/')
+        self.assertEqual(res.status_code, 200)
```
Comment (timgraham) [inline tests/generic_views/test_dates.py]: Seems like it would be useful to check the results.

File: tests/generic_views/models.py [hunk 7]
```diff
@@ -30,6 +30,14 @@ def __str__(self):
         return self.name
 
 
+class Series(models.Model):
+    name = models.CharField(max_length=100)
+    pubdate = models.DateField()
+
+    def __str__(self):
```
Comment (timgraham) [inline tests/generic_views/models.py]: Is this used? Can an existing model be reused? We try to avoid adding new models if possible since that complicates and slows the tests.

File: django/views/generic/dates.py [hunk 8]
```diff
@@ -687,7 +698,8 @@ def _get_next_prev(generic_view, date, is_previous, period):
         # Snag the first object from the queryset; if it doesn't exist that
         # means there's no next/previous link available.
         try:
-            result = getattr(qs[0], date_field)
+            result = functools.reduce(
```
Comment (timgraham) [inline django/views/generic/dates.py]: These lines aren't straightforward. Maybe there's an alternative without functools.reduce that would improve readability?  Even so, this could be a single line... we allow up to 119 characters when it improves readability.

File: django/views/generic/dates.py [hunk 9]
```diff
@@ -245,8 +247,17 @@ def uses_datetime_field(self):
...
         """
         model = self.get_queryset().model if self.model is None else self.model
-        field = model._meta.get_field(self.get_date_field())
-        return isinstance(field, models.DateTimeField)
+
+        pieces = self.get_date_field().split(LOOKUP_SEP)
+        fields = []
+        for piece in pieces:
+            parent = model
+            if (fields and fields[-1].is_relation and
```
Comment (timgraham) [inline django/views/generic/dates.py]: This code isn't very understandable. It seems a helper function, method, or something is needed.

File: django/views/generic/dates.py (hunk 2 above)
Comment (timgraham) [inline django/views/generic/dates.py]: Some helper or something is needed because the current code is quite difficult to read and understand.

**Comment by timgraham** [comment]
After you update a PR, uncheck "Patch needs improvement" on the ticket so that it returns to the review queue.

File: tests/generic_views/models.py (hunk 7 above)
Comment (lefterisnik) [inline tests/generic_views/models.py]: I changed the code and now it uses an existing one `Artist`

File: django/views/generic/dates.py (hunk 8 above)
Comment (lefterisnik) [inline django/views/generic/dates.py]: I think the `functool.reduce` suits nice in this case. Any other alternative will require more code, hence will add more complexity.

**Comment by lefterisnik** [comment]
About your comment:

```
Some helper or something is needed because the current code is quite difficult to read and understand.
```

I created a helper function under `db/models/utils.py`.

File: django/views/generic/dates.py [hunk 10]
```diff
@@ -696,7 +699,7 @@ def _get_next_prev(generic_view, date, is_previous, period):
         # Snag the first object from the queryset; if it doesn't exist that
         # means there's no next/previous link available.
         try:
-            result = getattr(qs[0], date_field)
+            result = functools.reduce(getattr, date_field.split(LOOKUP_SEP), qs[0])
```
Comment (carltongibson) [inline django/views/generic/dates.py]: The use of `reduce` here not obvious. I think it needs a comment, something like: 

    # Walk parts of `date_field` returning last element. 

What happens here if `date_field` includes an incorrect part, e.g. `artist__not_there`? It's not going to raise an `IndexError` in that case. Do we handle this at all?

File: django/db/models/utils.py [hunk 11]
```diff
@@ -19,3 +22,16 @@ def make_model_tuple(model):
             "Invalid model reference '%s'. String model references "
             "must be of the form 'app_label.ModelName'." % model
         )
+
+
+def get_field_from_path(model, path):
```
Comment (carltongibson) [inline django/db/models/utils.py]: This could do with a docstring, and tests I think. 

What happens if `path` contains an error? e.g. `"user__groups__not_there"` Do we need to handle that at all?

File: django/db/models/utils.py [hunk 12]
```diff
@@ -19,3 +22,16 @@ def make_model_tuple(model):
...
+    pieces = path.split(LOOKUP_SEP)
+    fields = []
+    for piece in pieces:
+        parent = model
+        if (fields and fields[-1].is_relation and
```
Comment (carltongibson) [inline django/db/models/utils.py]: There must be a clearer way to spell out the intent of this block. 

Thoughts: 

* Using `all()` with a comment for why for each element? (We can tidy that once it's clearer)
* Giving this `fields[-1].remote_field.model` a meaningful name would help.

File: django/db/models/utils.py (hunk 11 above)
Comment (timgraham) [reply django/db/models/utils.py]: Unless this function can be used elsewhere (In which case I would introduce it and use it in a separate PR), I would leave this as a private helper function in `views/generic/utils.py`

**Comment by felixxm** [comment]
Closing due to inactivity.
